File: Src/Icm.Core/IO/CompositeWriter.vb
Imports System.IO

Namespace Icm.IO

    ''' <summary>
    ''' TextWriter composed by other TextWriters. Every write
    ''' is transferred to all the TextWriters.
    ''' </summary>
    ''' <remarks>
    ''' </remarks>
    ''' <history>
    ''' 	[icalvo]	03/12/2005	Created
    ''' </history>
    Public Class CompositeWriter
        Inherits TextWriter

        Private ReadOnly _textWriters As New List(Of TextWriter)()

        Protected ReadOnly Property textWriters() As IList(Of TextWriter)
            Get
                Return _textWriters
            End Get
        End Property

        Public Sub New(ByVal ParamArray tws() As TextWriter)
            MyBase.New(CultureInfo.CurrentCulture)
            Add(tws)
        End Sub

        Public Shadows Sub Add(ByVal ParamArray tws() As TextWriter)
            For Each tw In tws
                If tw Is Nothing Then
                    Throw New ArgumentNullException("tws")
                End If
                textWriters.Add(tw)
            Next
        End Sub

        Public Overloads Overrides Sub Write(ByVal value As Char)
            For Each tw As TextWriter In textWriters
                tw.Write(value)
            Next
        End Sub

        Public Overloads Overrides Sub Write(ByVal value As String)
            For Each tw As TextWriter In textWriters
                tw.Write(value)
            Next
        End Sub

        Public Overrides Sub Flush()
            For Each tw As TextWriter In textWriters
                tw.Flush()
            Next
        End Sub

        Public Overrides Sub Close()
            For Each tw As TextWriter In textWriters
                tw.Close()
            Next
        End Sub

        Public Overloads Overrides Sub WriteLine(ByVal value As String)
            For Each tw As TextWriter In textWriters
                tw.WriteLine(value)
            Next
        End Sub

        Public Overrides ReadOnly Property Encoding() As System.Text.Encoding
            Get
                Return System.Text.Encoding.Default
            End Get
        End Property
    End Class

End Namespace

File: Src/Icm.Core/IO/CompositeWriter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;

namespace Icm.IO
{

	/// <summary>
	/// TextWriter composed by other TextWriters. Every write
	/// is transferred to all the TextWriters.
	/// </summary>
	/// <remarks>
	/// </remarks>
	/// <history>
	/// 	[icalvo]	03/12/2005	Created
	/// </history>
	public class CompositeWriter : TextWriter
	{


		private readonly List<TextWriter> _textWriters = new List<TextWriter>();
		protected IList<TextWriter> TextWriters => _textWriters;

	    public CompositeWriter(params TextWriter[] tws) : base(CultureInfo.CurrentCulture)
		{
			Add(tws);
		}

		public void Add(params TextWriter[] tws)
		{
			foreach (var tw in tws) {
				if (tw == null) {
					throw new ArgumentNullException(nameof(tws));
				}

				TextWriters.Add(tw);
			}
		}

		public override void Write(char value)
		{
			foreach (TextWriter tw in TextWriters) {
				tw.Write(value);
			}
		}

		public override void Write(string value)
		{
			foreach (TextWriter tw in TextWriters) {
				tw.Write(value);
			}
		}

		public override void Flush()
		{
			foreach (TextWriter tw in TextWriters) {
				tw.Flush();
			}
		}

		public override void Close()
		{
			foreach (TextWriter tw in TextWriters) {
				tw.Close();
			}
		}

		public override void WriteLine(string value)
		{
			foreach (TextWriter tw in TextWriters) {
				tw.WriteLine(value);
			}
		}

		public override System.Text.Encoding Encoding {
			get {
				throw new InvalidOperationException("CompositeWriter does not have a single encoding because it is composed of several child TextWriters.");
			}
		}
	}

}